src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs

```diff
@@ -522,7 +492,67 @@ protected override PredicateValueKind SetValueForIsNullComparisonOperator(IOpera
                 if (inferInTargetAnalysisData)
                 {
                     // Set value for the CurrentAnalysisData.
-                    SetAbstractValueFromPredicate(key, value, defaultPointsToValueGenerator, sourceAnalysisData, targetAnalysisData);
+                    SetAbstractValueFromPredicate(key, value, targetType, defaultPointsToValueGenerator, compilation, sourceAnalysisData, targetAnalysisData);
+                }
+
+                return;
+
+                static void SetAbstractValueFromPredicate(
+                    AnalysisEntity analysisEntity,
+                    NullAbstractValue nullState,
+                    ITypeSymbol targetType,
+                    DefaultPointsToValueGenerator defaultPointsToValueGenerator,
+                    Compilation compilation,
+                    PointsToAnalysisData sourceAnalysisData,
+                    PointsToAnalysisData targetAnalysisData)
+                {
+                    AssertValidPointsToAnalysisData(sourceAnalysisData);
+                    AssertValidPointsToAnalysisData(targetAnalysisData);
+
+                    Debug.Assert(IsValidValueForPredicateAnalysis(nullState) || nullState == NullAbstractValue.Invalid);
+
+                    // Ensure that the predicated 'value' can be flowed from the target type to the analysis entity.
+                    if (!SymbolEqualityComparer.Default.Equals(targetType, analysisEntity.Type))
+                    {
+                        var conversion = compilation.ClassifyCommonConversion(targetType, analysisEntity.Type);
+                        if (!conversion.Exists)
+                        {
+                            // No conversion exists, so we bail out from flowing the predicated value.
+                            return;
+                        }
+
+                        if (!conversion.IsIdentity && !conversion.IsNumeric && !conversion.IsNullable)
+                        {
+                            // For 'Null' predicated value, there needs to be an explicit conversion
+                            // from targetType to the analysisEntity's type to flow the predicated value.
+                            // For 'NotNull' predicated value, there needs to be an implicit conversion
+                            // from targetType to the analysisEntity's type to flow the predicated value.
+                            if (nullState == NullAbstractValue.Null && conversion.IsImplicit ||
+                                nullState == NullAbstractValue.NotNull && !conversion.IsImplicit)
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    if (!sourceAnalysisData.TryGetValue(analysisEntity, out var existingValue))
+                    {
+                        existingValue = defaultPointsToValueGenerator.GetOrCreateDefaultValue(analysisEntity);
+                    }
```

Comment:
I don't understand this area well, but wanted to make sure whether we want to add the `analysisEntity` and `existingValue` pair to `sourceAnalysisData` or not.

Reply:
This method is only designed to modify `targetAnalysisData`...